From 628d2adbbecf9966abea3615bc2ebfb4c1a8d9be Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 29 Aug 2018 10:20:48 -0700
Subject: [PATCH] MANTA-3125 deliver libmanta via npm registry instead of git
 repo (setup for npm publishing; 'make cutarelease' target) Reviewed by: Jared
 Morrow <jm@joyent.com> Approved by: Jared Morrow <jm@joyent.com>

---
 Makefile     | 22 +++++++++++++++++++++-
 README.md    | 19 +++++++++++++++++++
 package.json |  1 -
 3 files changed, 40 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index bb0994f..48d0b38 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -62,6 +62,26 @@ CLEAN_FILES += node_modules
 test: all
 	$(NODEUNIT) test/*.test.js
 
+.PHONY: cutarelease
+cutarelease:
+	[[ -z `git status --short` ]]  # If this fails, the working dir is dirty.
+	@which json 2>/dev/null 1>/dev/null && \
+	    ver=$(shell json -f package.json version) && \
+	    name=$(shell json -f package.json name) && \
+	    publishedVer=$(shell npm view -j $(shell json -f package.json name)@$(shell json -f package.json version) version 2>/dev/null) && \
+	    if [[ -n "$$publishedVer" ]]; then \
+		echo "error: $$name@$$ver is already published to npm"; \
+		exit 1; \
+	    fi && \
+	    echo "** Are you sure you want to tag and publish $$name@$$ver to npm?" && \
+	    echo "** Enter to continue, Ctrl+C to abort." && \
+	    read
+	ver=$(shell cat package.json | json version) && \
+	    date=$(shell date -u "+%Y-%m-%d") && \
+	    git tag -a "v$$ver" -m "version $$ver ($$date)" && \
+	    git push --tags origin && \
+	    npm publish
+
 include ./tools/mk/Makefile.deps
 include ./tools/mk/Makefile.node_deps.targ
 include ./tools/mk/Makefile.targ
diff --git a/README.md b/README.md
index 25d3a79..53b3c45 100644
--- a/README.md
+++ b/README.md
@@ -32,3 +32,22 @@ $ export ELECTRIC_MORAY=electric-moray.emy-10.joyent.us
 $ export MAHI_HOST=authcache.emy-10.joyent.us
 $ make prepush
 ```
+
+# Release process
+
+Manta components that use libmanta should use versions published to npm.
+To release a new libmanta:
+
+1. Make a commit to set the intended version in "package.json#version".
+
+2. Get that commit approved and merged via <https://cr.joyent.us>, as with all
+   commits to this repo. See the discussion of contribution at the top of this
+   readme.
+
+3. Once that is merged and you've updated your local copy, run:
+
+    ```
+    make cutarelease
+    ```
+
+   This will ensure a clean working copy, git tag, and npm publish.
diff --git a/package.json b/package.json
index 1f4b014..8d0878e 100644
--- a/package.json
+++ b/package.json
@@ -3,7 +3,6 @@
     "description": "common functions et al for manta services",
     "version": "1.1.1",
     "author": "Joyent (joyent.com)",
-    "private": true,
     "main": "./lib/index.js",
     "repository": {
         "type": "git",
-- 
2.21.0

